src/lib/viz/legend/LegendWidget.ts

```diff
@@ -0,0 +1,57 @@
+import { Layer } from '@/viz/layer';
+import { queryDOMElement } from '@/core/utils/dom';
+import { CartoError } from '@/core/errors/CartoError';
+
+export type LegendGeometryType = 'point' | 'line' | 'polygon';
+
+export type LegendProperties = {
+  type: LegendGeometryType;
+  color?: string;
+  marker?: string;
+  strokeColor?: string;
+  strokeStyle?: string;
+  label?: string;
+  width?: number;
+};
+
+export class LegendWidget {
+  constructor(element: string | HTMLElement, layer: Layer, options = {}) {
+    const domElement = queryDOMElement(element);
+    validateParameters(domElement, layer, options);
+    const legendWidget = domElement as any;
+
+    if (layer.isReady()) {
+      legendWidget.data = layer.getLegendData(options);
+    } else {
+      layer.on('dataReady', () => {
```

Comment:
Let's try to use internally our enums for event names, in this case `LayerEvent.DATA_READY`, currently inside of Layer.ts

Reply:
Done!